keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -129,27 +143,49 @@ use un-encrypted...
+
 ## Proposal
 
-For alpha the kubelet will keep a list, across reboots of host and restart of
-kubelet, of container images that required authentication and a list of the
-authentications that successfully pulled the image.
-For beta an API will be considered to manage the ensure metadata.
+The kubelet will track container images and the list of authentication information
+that lead to their successful pulls. This data will be persisted across reboots
+of the host and restarts of the kubelet.
 
-The kubelet will ensure any image in the list is always pulled if an authentication
-used is not present, thus enforcing authentication / re-authentication.
+The persistent authentication data storage will be done using files kept in the
+kubelet directory on the node. The content of these files will be structured and
+versioned using standard config file API versioning techniques.
 
-There will be two different kubelet configuration options added, as well as a feature
-gate to gate their use:
-- `pullImageSecretRecheck`
-    - A boolean that toggles this behavior. If `false`, the kubelet will fallback to the
-      old behavior: only pull an image if it's not present.
-- `pullImageSecretRecheckPeriod`
-    - the period after which the kubelet's cache will be invalidated,
-      thus causing rechecks for all `IfNotPresent` images that are recreated.
-    - If set to `0s`, or `0`, but `pullImageSecretRecheck` is `true`, then
-      the kubelet will never invalidate its cache, but will maintain one.
+The kubelet will ensure any image requiring credential verification is always pulled if authentication information
+from an image pull is not yet present, thus enforcing authentication / re-authentication.
 
+Two new kubelet configuration options will be added, as well as a feature
+gate to gate their use:
+- `imagePullCredentialsVerificationPolicy`
+    - An enum of 4 possible values:
+      - `NeverVerify`
+        - **original behavior before this feature**
+        - anyone on a node can use any image present on the node
+      - `NeverVerifyNodePreloadedImages`
+        - **default value**
+        - images that were pulled to the node by something else than the kubelet can be used without reverifying pull credentials
+      - `NeverVerifyAllowlistedImages`
+        - like `NeverVerifyNodePreloadedImages` but only node images from `preloadedImagesVerificationAllowlist` don't require reverification
+      - `AlwaysVerify`
+        - all images pulled with credentials require credential reverification
+- `preloadedImagesVerificationAllowlist`
+  - a string list to be used with the `NeverVerifyAllowlistedImages` image pull credentials verification policy
+  - images specified by this list will not require credential reverification if they are requested
+    by a pod with `IfNotPresent` or `Never` image pull policies
+  - the list accepts wildcard suffix in a string where `*` is the wildcard character
```

Comment:
I also realized some registries have leading regionalized subdomains they might want to wildcard (`foo.example.com/...`, etc). I'm on the fence about including that or not in the initial revision, but I wouldn't object to simple prefix/suffix glob. Might be worth checking main registries to see if that is widespread

Reply:
It looks like regionalized subdomains are quite common:
- https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview
- https://docs.oracle.com/en-us/iaas/Content/Registry/Concepts/registryprerequisites.htm
- https://cloud.google.com/artifact-registry/docs/transition/gcr-repositories#gcr-domain-support
- https://learn.microsoft.com/en-us/azure/container-registry/container-registry-geo-replication#example-use-case

I very briefly read the above documents. It seems to me that the use of regional registry is usually deliberate, and so maybe it should be in our config, too.

There was also a mention of infixed regional domains in the IBM document - see the "Private domain name" column in https://cloud.ibm.com/docs/Registry?topic=Registry-registry_overview#registry_regions_local

I'd suggest starting with suffixes for the first alpha, we can always modify the behavior to be looser.